SAMPLE_FORMAT.md

```diff
@@ -25,19 +57,128 @@ specifically mentioned in the format, samples should make best-effort attempts i
+[Gradle](https://gradle.org/) support is optional as we don't yet have regular testing.
+
+## Testing
+Snippets should have integration tests that should verify the snippet works and compiles correctly.
+Creating mocks for these tests are optional. These tests should capture output created by the
+snippet to verify that it works correctly. See the tests in the canonical for an example of how to
+do this correctly. 
+
+* Test Library: [JUnit4](https://junit.org/junit4/)
+* Test Runner: [Maven Failsafe plugin](https://maven.apache.org/surefire/maven-failsafe-plugin/)
+and [Maven Surefire plugin](https://maven.apache.org/surefire/maven-surefire-plugin/).
 
+As an example, the following test code shows how we test a region tag called `region_tag`:
+```java
+package com.google.example;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Store region_tag in test method name, camel-cased
+ */ 
+@RunWith(JUnit4.class)
+public class SomeClassTest {
+  @Test public void regionTag_shouldPass() throws Exception { }
+
+  @Test public void regionTag_shouldFail() throws Exception {
+    Assert.fail("should fail");
+  }
+}
+```
+You will note that the "_" in `region_tag` is removed, and "_" is used to separate regionTags from
+test descriptions.
+
+It is also possible to use annotations to provide info for `region_tag` if you need to do this,
+please contact one of the repo admins.
+
+### Testing Setup
+Most samples require a GCP project and billing account. Keep the following in
+mind when setting up tests.
+
+* **Environment variables**  
+  Minimize additional environment variables that need to be set to run the tests.
+  If you do require additional environment variables, they should be added to
+  `run_tests.sh`.
+
+  Existing environment variables include:
+  * `GOOGLE_APPLICATION_CREDENTIALS`
+  * `GOOGLE_CLOUD_PROJECT`
+  * `PROJECT_ID`
+
+* **API library**  
+  If an API needs to be enabled in the testing project, add this information to the
+  directory's README and to the comments in the PR. If there is no README.md file, add one
+  in your PR.
+
+* **IAM**
+  Some API's require that the service account have some additional capibilities, these should also
+  be mentioned in both the PR and the README.
+
+* **Cloud resources**  
+  Most Java samples create the Cloud resources that they need to run. If this
+  is resource intensive or not possible, add instructions to the directory's CONTRIBUTING.md file
+  to add the resource to the testing project.
+
+* **Keys and Secrets**
+  Add a note in the pull request, in order for a Java maintainer to assist you
+  in adding keys and secrets to the testing project.
+  
+### Running Locally
+Run tests locally with commands:
+
+* Maven: `mvn verify`
+* Gradle: `gradle build test`
+
+### Gradle Specifics
+#### NEEDS WORK
```

Comment:
Will this be updated in another PR?

Reply:
Yea - but we need to agree on a canonical build file, and add testing.  I didn't want to remove it, so I added that info.